tests/utils/runner.js

```diff
@@ -0,0 +1,60 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+const { spawn, execSync } = require("child_process");
+const { join } = require("path");
+
+process.chdir(join(__filename, "../.."));
+
+console.log("Bundling Tabster...");
+execSync("npx vite build --config vite.config.js");
+console.log("Bundling Tabster done.");
+
+console.log("Serving test page...");
+const serveArgs = ["vite", "preview", "--config", "vite.config.js"];
+const port = parseInt(process.env.PORT || "0", 10);
+if (port > 0) {
+    serveArgs.push("--port", port);
+}
+const serve = spawn("npx", serveArgs);
+
+let testRun;
+
+serve.stdout.on("data", (data) => {
+    if (testRun) {
+        return;
+    }
+
+    const match = data.toString().match(/http:\/\/localhost:(\d+)\//);
+
+    if (match) {
+        let port = parseInt(match[1], 10);
+
+        if (!(port > 0)) {
+            port = 8080;
+        }
+
+        console.log(`Serving test page on port ${port}`);
+
+        testRun = spawn("jest", process.argv.slice(2), {
```

Comment:
what you're doing here is creating a custom bit of magic on top of jest, which is generally an antipattern. I would suggest creating a custom jest environment for tabster, which would be far more intuitive as a reusable thing.

https://jestjs.io/docs/configuration#testenvironment-string

Reply:
Sure, I'll look into that. This custom bit of magic is a quick way to test if the idea works in general.